File: outco/matrix/wordSearch.js
/* 
314 - Word Search

Successfully tested the function 1/22/23.

Complexity Analysis:
Where M = matrix height, N = matrix width, L = length of word
Time Complexity: O(MN * 4^L)
Auxiliary Space Complexity: O(MN)

Given a matrix of characters a-z and a word,
return true or false about whether the word can be found within the matrix. 
Words can only be built from adjacent letters (up, down, left, right), but not diagonals.
Same elements cannot be used multiple times in the same word.

Input:

[['a', 'b', 'c', 'd'],
 ['e', 'f', 'g', 'h'],
 ['i', 'd', 'o', 'j'],
 ['k', 'l', 'm', 'n']]

'dog'

Output: true
loop until finding the first letter
pass coordinates to traverse helper method
helper method checks ( x + 1, y ) (right), ( x - 1, y) (left), ( x, y + 1 ) (down), ( x , y - 1 ) (up)
for second letter...all remaining letters recursively

Terminate traversal: 
- if letter is already visited (set to null)
- if the next spot is out of bounds
- if no more letters left to search for 
- if the letter is the last letter in the word (return true)                    

recursive tree
                   'd'
            /     /   \    \
          'o'   'i'   'l'  'f'
       select o...


*/

const wordSearch = (matrix, word) => {
    
    let height = matrix.length;
    let width = matrix[0].length;
    let xCoordinate = 0;
    let yCoordinate = 0;
    let output;

    for (let x = 0; x < height; x++){
        for (let y = 0; y < width; y++){
            if (matrix[x][y] === word[0]){
                xCoordinate = x;
                yCoordinate = y;
            }
        }
    }
//     Terminate traversal: 
// - if letter is already visited (set to null)
// - if the next spot is out of bounds
// - if no more letters left to search for 
// - if the letter is the last letter in the word (return true)                    

    const traverse = (m, n, word, index) => {
      
        if ( m < 0 || n < 0 || m > height || n > width){
            return;
        }

        if (matrix[m][n] === null){
            return;
        }
        if (index === word.length - 1 && matrix[m][n] === word[word.length -1]){
            output = true;
            return;
        }
        
        if (matrix[m][n] === word[index]){
            let temp = matrix[m][n]
            matrix[m][n] = null;
            traverse(m - 1, n, word, index + 1);
            traverse(m + 1, n, word, index + 1);
            traverse(m, n - 1, word, index + 1);
            traverse(m, n + 1, word, index + 1);
            matrix[m][n] = temp;
        }
    }
    traverse(xCoordinate, yCoordinate, word, 0)
    return output === undefined ? false : output
}


console.log(wordSearch([['a', 'b', 'c', 'd'],
                        ['e', 'f', 'g', 'h'],
                        ['i', 'd', 'o', 'j'],
                        ['k', 'l', 'm', 'n']], 'dog')) // true

console.log(wordSearch([['t', 'a', 'c', 'd'],
                        ['e', 'f', 'g', 'h'],
                        ['i', 'd', 'o', 'j'],
                        ['k', 'l', 'm', 'n']], 'cat')) // true

console.log(wordSearch([['t']], 't')) // true

console.log(wordSearch([['t', 'a', 'c', 'd'],
                         ['e', 'f', 'g', 'h'],
                         ['i', 'd', 'o', 'j'],
                         ['k', 'l', 'm', 'n']], '1123132134')) // false

File: c++journey/outco_in_cpp/backtracking/wordSearch.cpp
#include <array>
#include <cmath>
#include <iterator>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <iostream>
#include <array>
#include <set>
using namespace std;
/* 
314 - Word Search

in C++

Complexity Analysis:
Where M = matrix height, N = matrix width, L = length of word
Time Complexity: O(MN * 4^L)
Auxiliary Space Complexity: O(MN)

Given a matrix of characters a-z and a word,
return true or false about whether the word can be found within the matrix. 
Words can only be built from adjacent letters (up, down, left, right), but not diagonals.
Same elements cannot be used multiple times in the same word.

Input:

[['a', 'b', 'c', 'd'],
 ['e', 'f', 'g', 'h'],
 ['i', 'd', 'o', 'j'],
 ['k', 'l', 'm', 'n']]

'dog'

Output: true
loop until finding the first letter
pass coordinates to traverse helper method
helper method checks ( x + 1, y ) (right), ( x - 1, y) (left), ( x, y + 1 ) (down), ( x , y - 1 ) (up)
for second letter...all remaining letters recursively

Terminate traversal: 
- if letter is already visited (set to null)
- if the next spot is out of bounds
- if no more letters left to search for 
- if the letter is the last letter in the word (return true)                    

recursive tree
                   'd'
            /     /   \    \
          'o'   'i'   'l'  'f'
       select o...


*/
    bool traverse(vector<vector<char> >& matrix, const string& wordStr,
                int mInt, int nInt, int indexInt, int heightInt, int widthInt,
                unordered_set<string>& visitedSet){
        string keyStr = to_string(mInt) + "_" + to_string(nInt);
        if (mInt < 0 || nInt < 0 || mInt >= heightInt || nInt >= widthInt){
            return false;
        }
        if (matrix[mInt][nInt] == '\0' || matrix[mInt][nInt] != wordStr[indexInt] || visitedSet.count(keyStr) > 0){
            return false;
        }
        if (indexInt == wordStr.length() - 1 && matrix[mInt][nInt] == wordStr[wordStr.length() - 1]){
            return true;
        }
        visitedSet.insert(keyStr);
        bool upToF = traverse(matrix, wordStr, mInt - 1, nInt, indexInt + 1, heightInt, widthInt, visitedSet);
        bool downToF = traverse(matrix, wordStr, mInt + 1, nInt, indexInt + 1, heightInt, widthInt, visitedSet);
        bool leftToF = traverse(matrix, wordStr, mInt, nInt - 1, indexInt + 1, heightInt, widthInt, visitedSet);
        bool rightToF = traverse(matrix, wordStr, mInt, nInt + 1, indexInt + 1, heightInt, widthInt, visitedSet);
        visitedSet.erase(keyStr);
        return upToF || downToF || leftToF || rightToF;
    }

    bool wordSearch(vector<vector <char> >& matrix, const string& wordStr){
        int heightInt = matrix.size();
        int widthInt = matrix[0].size();
        unordered_set<string> visitedSet;
        for (int xInt = 0; xInt < heightInt; xInt++){
            for (int yInt = 0; yInt < widthInt; yInt++){
                if (matrix[xInt][yInt] == wordStr[0] && traverse(matrix, wordStr, xInt, yInt, 0, heightInt, widthInt, visitedSet)){
                    return true;
                }
            }
        }
        return false;
    }

    string converToFToString(bool ToF){
        if (ToF == 0){
            return "false";
        }
        if (ToF == 1){
            return "true";
        }
        return "";
    }
    int main(){
        vector<vector<char> > matrix1 = {
            {'A', 'B', 'C', 'E'},
            {'S', 'F', 'C', 'S'},
            {'A', 'D', 'E', 'E'}
        };
        string wordStr1 = "ABCB";
        bool resultToF1 = wordSearch(matrix1, wordStr1);
        string resultStr1 = converToFToString(resultToF1);
        cout << "Result 1: " << resultStr1 << endl;
        return 0;
    }
